site/app/controllers/ChatroomController.php

```diff
@@ -0,0 +1,234 @@
+        $active_chatrooms = $repo->findBy(['is_active' => true], ['id' => 'ASC']);
+
+        if ($user->accessAdmin()) {
+            return new WebResponse(
+                'Chatroom',
+                'showChatPageInstructor',
+                $chatrooms
+            );
+        }
+        else { // Student view
+            return new WebResponse(
+                'Chatroom',
+                'showChatPageStudent',
+                $active_chatrooms
+            );
+        }
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/newChatroom", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function addChatroom(): RedirectResponse {
+        $em = $this->core->getCourseEntityManager();
+
+        $chatroom = new Chatroom();
+        $chatroom->setTitle($_POST['title']);
+        $chatroom->setDescription($_POST['description']);
+        $chatroom->setHostId($this->core->getUser()->getId());
+        $chatroom->setHostName($this->core->getUser()->getDisplayFullName());
+
+        $em->persist($chatroom);
+        $em->flush();
+
+        $this->core->addSuccessMessage("Chatroom successfully added");
+        return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}", methods={"GET"})
+     * @param string $chatroom_id
+     * @return RedirectResponse|WebResponse
+     */
+     public function getChatroom(string $chatroom_id): WebResponse|RedirectResponse {
+         if (!is_numeric($chatroom_id)) {
+            $this->core->addErrorMessage("Invalid Chatroom ID");
+            return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+        }
+
+        $repo = $this->core->getCourseEntityManager()->getRepository(Chatroom::class);
+        $chatroom = $repo->find($chatroom_id);
+
+        return new WebResponse(
+            'Chatroom',
+            'showChatroom',
+            $chatroom,
+        );
+     }
+
+     /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/anonymous", methods={"GET"})
+     * @param string $chatroom_id
+     * @return RedirectResponse|WebResponse
+     */
+     public function getChatroomAnon(string $chatroom_id): WebResponse|RedirectResponse {
+         if (!is_numeric($chatroom_id)) {
+            $this->core->addErrorMessage("Invalid Chatroom ID");
+            return new RedirectResponse($this->core->buildCourseUrl(['chat']));
+        }
+
+        $repo = $this->core->getCourseEntityManager()->getRepository(Chatroom::class);
+        $chatroom = $repo->find($chatroom_id);
+
+        return new WebResponse(
+            'Chatroom',
+            'showChatroom',
+            $chatroom, true,
+        );
+     }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/deleteChatroom", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function deleteChatroom(): JsonResponse {
```

Comment:
Should we have the ability to delete chatrooms?  Perhaps we should keep them around but add a boolean "deleted" column in case there's ever a desire to dig up the chatroom in the future.

Reply:
I thought of this before. I added a column in the database table to allow a chat room to be "session" based. Toggle on a chat room will allow it to be active and available to the students, and stopping the session will take away students' ability to type into the chat and/or entering the chat room. So instead of a "deleted" column, a chat room would just be "inactive", and the host and activate it again easily if they wish to dig it up or reuse it.

On the other hand if a chat room is deleted it is deleted, so there's no confusion.